File: ra_slam/custom_factors/SESyncFactor3d.py
import gtsam
import numpy as np
from typing import List, Optional

# Generators for the tangent space of SO(3)
G1 = np.array([[0, 0, 0], [0, 0, -1], [0, 1, 0]])

G2 = np.array([[0, 0, 1], [0, 0, 0], [-1, 0, 0]])

G3 = np.array([[0, -1, 0], [1, 0, 0], [0, 0, 0]])


def tangent_matrix(omega: np.array):
    assert len(omega) == 3
    Omega = omega[0] * G1 + omega[1] * G2 + omega[2] * G3
    return Omega


class RelativePose3dFactor(gtsam.CustomFactor):
    def __init__(
        self,
        key_i: int,
        key_j: int,
        relative_rotation: np.ndarray,
        relative_translation: np.ndarray,
        rotation_precision: float,
        translation_precision: float,
        # noiseModel: gtsam.noiseModel.Base,
    ) -> None:
        noise_model = gtsam.noiseModel.Diagonal.Precisions(
            2 * np.array([translation_precision] * 3 + [rotation_precision] * 9)
        )
        # noise_model = noiseModel
        super().__init__(noise_model, [key_i, key_j], self.relative_pose3d_error_func)
        self._key_i = key_i
        self._key_j = key_j
        self._relative_rotation = relative_rotation.reshape((3, 3))
        self._relative_translation = relative_translation.reshape((3, 1))

    def relative_pose3d_error_func(
        self,
        this: gtsam.CustomFactor,
        values: gtsam.Values,
        jacobians: Optional[List[np.ndarray]],
    ) -> float:
        """Error function for RelativePose3dFactor with Langevin noise on the
        rotation and Gaussian noise on the translation.
        Args:
            measurement (np.ndarray): The measurement, a 3x1 vector of [x, y, theta]
            this (gtsam.CustomFactor): The factor itself
            values (gtsam.Values): The current estimate of the variables
            jacobians (Optional[List[np.ndarray]]): The jacobians of the error function
        Returns:
            float: The unwhitened error
        """
        # Get the keys
        key1 = this.keys()[0]
        key2 = this.keys()[1]

        # Get the values
        Xi = values.atPose3(key1)
        Xj = values.atPose3(key2)

        # Get the measurement
        measured_translation = self._relative_translation
        measured_rotation = self._relative_rotation

        # Get the error
        residual = np.zeros(12)
        # print(Xi.rotation().matrix())
        # print(measured_translation)
        trans_error = (Xj.translation() - Xi.translation()).reshape((3, 1)) - (
            Xi.rotation().matrix() @ measured_translation
        )
        # print(type(measured_rotation))
        rot_error = Xj.rotation().matrix() - (
            Xi.rotation().matrix() @ measured_rotation
        )
        residual[0:3] = trans_error[:, 0]
        residual[3:12] = rot_error.reshape(9)

        # Get the jacobians

        #  J = [ dt_err / dR, dt_err / dt ;
        #   dvec(R_err) / dR, 0];
        if jacobians:
            # build the first jacobian
            H1 = np.zeros((12, 6))
            Ri = Xi.rotation().matrix()
            # dRi_domega = tangent_matrix(1.0)

            # Set dt_err / dR_i
            H1[0:3, 0:3] = Ri @ tangent_matrix(measured_translation[:, 0])

            # Set dt_err / dt
            H1[0:3, 3:6] = -Ri
            # H1->topRightCorner<2, 1>() = -Ri * dRi_domegai * tij;
            # H1[0:3, 3] = (-Ri @ G1 @ measured_translation)[:,0]
            # H1[0:3, 4] = (-Ri @ G2 @ measured_translation)[:,0]
            # H1[0:3, 5] = (-Ri @ G3 @ measured_translation)[:,0]

            # dR_err / dt_i = 0

            # Set dR_err / dR_i
            # Eigen::Matrix2d dRerr_dRi = -Ri * dRi_domegai * Rij.matrix();
            # dRerr_dRi = -Ri @ dRi_domega @ measured_rotation
            dRerr_dRiG1 = -Ri @ G1 @ measured_rotation
            dRerr_dRiG2 = -Ri @ G2 @ measured_rotation
            dRerr_dRiG3 = -Ri @ G3 @ measured_rotation

            # H1->bottomRightCorner<4, 1>() =
            #     Eigen::Map<Eigen::Vector4d>(dRerr_dRi.data());
            H1[3:12, 0] = dRerr_dRiG1.reshape(9)
            H1[3:12, 1] = dRerr_dRiG2.reshape(9)
            H1[3:12, 2] = dRerr_dRiG3.reshape(9)

            jacobians[0] = H1

            # build the second jacobian
            H2 = np.zeros((12, 6))
            Rj = Xj.rotation().matrix()
            # dRj_domega = tangent_matrix(1.0)

            # Set dt_err / dt_j
            # H2->topLeftCorner<2, 2>() = Rj;
            H2[0:3, 3:6] = Rj

            # dt_err / dR_j = 0
            # dR_err / dt_j = 0

            # dR_err / dR_j
            # Eigen::Matrix2d dRerr_dRj = Rj * tangent_matrix(1);
            # dRerr_dRj = Rj @ dRj_domega
            dRerr_dRjG1 = Rj @ G1  # @ measured_rotation
            dRerr_dRjG2 = Rj @ G2  # @ measured_rotation
            dRerr_dRjG3 = Rj @ G3  # @ measured_rotation

            # H2->bottomRightCorner<4, 1>() =
            #     Eigen::Map<Eigen::Vector4d>(dRerr_dRj.data());
            H2[3:12, 0] = dRerr_dRjG1.reshape(9)
            H2[3:12, 1] = dRerr_dRjG2.reshape(9)
            H2[3:12, 2] = dRerr_dRjG3.reshape(9)

            jacobians[1] = H2

        # Return the error
        return residual


def theta2matrix(theta: float) -> np.ndarray:
    """Converts a rotation angle to a rotation matrix.
    Args:
        theta (float): The rotation angle
    Returns:
        np.ndarray: The rotation matrix
    """
    return np.array([[np.cos(theta), -np.sin(theta)], [np.sin(theta), np.cos(theta)]])


def uniform_random_vec3D():
    # See J. Arvo "Fast Random Rotation Matrices" (doi=10.1.1.53.1357)
    # for uniformly random unit vector computation
    x1 = np.random.uniform()
    x2 = np.random.uniform()
    vhat = np.array(
        [
            np.cos(2.0 * np.pi * x1) * np.sqrt(x2),
            np.sin(2.0 * np.pi * x1) * np.sqrt(x2),
            np.sqrt(1.0 - x2),
        ]
    )
    return np.mat(vhat).T


def sample_uniform_rotation(d):
    """
    A uniform sampler for SO(d), d=2,3

    See J. Arvo "Fast Random Rotation Matrices" (doi=10.1.1.53.1357) for uniformly
    random unit vector computation

    """
    x1 = np.random.uniform()
    theta = 2.0 * np.pi * x1
    rot2D = theta2matrix(theta)
    if d == 2:
        return rot2D
    else:
        # d == 3
        R = np.eye(3)
        R[:2, :2] = rot2D
        v = uniform_random_vec3D()
        vdv = v.dot(v.T)
        H = np.eye(3) - 2.0 * v.dot(v.T)
        return -H.dot(R)


if __name__ == "__main__":
    measured_translation = np.array([1.0, 2.0, 3.0])
    measured_rotation = sample_uniform_rotation(3)

    theta_1 = 0.0
    pose1 = gtsam.Pose3(r=gtsam.Rot3(np.eye(3)), t=np.zeros(3))

    translation_precision = 1
    rotation_precision = 1

    noise_model = gtsam.noiseModel.Diagonal.Precisions(
        2 * np.array([translation_precision] * 3 + [rotation_precision] * 9)
    )

    robustModel = gtsam.noiseModel.Robust.Create(
        gtsam.noiseModel.mEstimator.Cauchy.Create(1), noise_model
    )

    for i in range(10):
        tran_2 = np.array([1.0, 2.0, 3.0])  # np.random.uniform(-1, 1, 3)
        rot_2 = sample_uniform_rotation(3)
        pose2 = gtsam.Pose3(r=gtsam.Rot3(rot_2), t=tran_2)
        symbol1 = gtsam.symbol("x", 1)
        symbol2 = gtsam.symbol("x", 2)

        rot_precision = 1000
        tran_precision = 100
        factor = RelativePose3dFactor(
            symbol1,
            symbol2,
            measured_rotation,
            measured_translation,
            rot_precision,
            tran_precision,
        )

        values = gtsam.Values()
        values.insert(symbol1, pose1)
        values.insert(symbol2, pose2)

        error = factor.error(values)
        linearized_factor = factor.linearize(values)
        print(linearized_factor)

        # get the error and the jacobians
        jacobians = [np.zeros((6, 3)), np.zeros((6, 3))]
        error = factor.error(values)

        print(f"Error: {error}")

        # manually calculate the Langevin error
        # rot_error = rot2 - rot1 @ measured_rotation
        rot1 = pose1.rotation().matrix()
        rot2 = pose2.rotation().matrix()
        rot_error = rot2 - rot1 @ measured_rotation
        rot_error_norm = np.linalg.norm(rot_error, "fro") ** 2
        print(f"Rot error should be (less than with cauchy): {rot_error_norm}")

File: ra_slam/custom_factors/SESyncFactor3d.cpp
#include "SESyncFactor3d.h"

Eigen::Matrix3d cross_product_matrix(const Eigen::Vector3d& omega)
{
    Eigen::Matrix3d S;
    S << 0, -omega(2), omega(1),
         omega(2), 0, -omega(0),
         -omega(1), omega(0), 0;

            return S;
}

gtsam::Vector SESyncFactor3d::evaluateError(const gtsam::Pose3& Xi, const gtsam::Pose3& Xj, boost::optional<gtsam::Matrix&> H1, boost::optional<gtsam::Matrix&> H2) const
{
    gtsam::Vector residual(12);

    // First, compute the translational component of the residual vector
    residual.head<3>() = (Xj.translation() - Xi.translation() - Xi.rotation() * tij);

    // Next, compute the rotational component of the residual vector
    gtsam::Matrix Rerr_mat = Xj.rotation().matrix() - (Xi.rotation() * Rij).matrix();
    residual.tail<9>() = Eigen::Map<Eigen::VectorXd>(Rerr_mat.data(), 9);


    /** For derivative computations, note that GTSAM considers Pose3 tangent vectors to be laid out as (omega, v) in R^6, where omega is angular velocity and v translational velocity).
     *
     * Therefore, the Jacobian of the residual vector
     *
     * r = [tj - ti - Ri tij;
     *      vec(Rj - Ri * Rij)]
     *
     * has the block structure
     *
     * J = [ dt_err / dR, dt_err / dt ;
     *       dvec(R_err) / dR, 0];
     */

    if(H1)
    {
        // Compute the Jacobian of the residual vector wrt pose Xi.  Note that this will be a 12 x 6 matrix, since the residual vector is 12-dimensional, but the intrinsic dimension of Xi is 6.
        *H1 = Eigen::MatrixXd::Zero(12, 6);

        /** Set dt_err / dR_i
         *
         * Note that since t_err = t_j - t_i - R_i t_ij is linear in R_i, then
         *
         * dt_err / dw = - d / dw [R_i exp([w]_x)] t_ij = - R_i (d / dw) ( [w]_x t_ij ) =  R_i d/dw ([t_ij]_x w) = R_i [t_ij]_x
         *
         */

        Eigen::MatrixXd Ri = Xi.rotation().matrix();
        H1->topLeftCorner<3,3>() = Ri * cross_product_matrix(tij);

        // Set dt_err / dt_i
        H1->topRightCorner<3,3>() = -Ri;



        // Set dvec(R_err) / dR
        Eigen::Matrix<double,9,3> dvecRerr_dRi = Eigen::Matrix<double,9,3>::Zero();

        for(unsigned int c = 0; c < 3; c++)
            for(unsigned int r = 0; r < 3; r++)
                dvecRerr_dRi.row(3*c + r) = -Rij.column(c + 1).cross(Ri.row(r));

        H1->bottomLeftCorner<9,3>() = dvecRerr_dRi;

        // dvec(R_err) / dt_i = 0
    }

    if(H2)
    {
        // Compute the Jacobian of the residual vector wrt pose Xi.  Note that this will be a 12 x 6 matrix, since the residual vector is 12-dimensional, but the intrinsic dimension of Xi is 6.
        *H2 = Eigen::MatrixXd::Zero(12, 6);

        Eigen::MatrixXd Rj = Xj.rotation().matrix();

        // dt_err / dRj = 0

        // Set dt_err / dtj
        H2->topRightCorner<3,3>() = Rj;


        // Set dvec(R_err) / dRj
        H2->block<3,1>(3, 1) = -Rj.col(2);
        H2->block<3,1>(3, 2) = Rj.col(1);

        H2->block<3,1>(6, 0) = Rj.col(2);
        H2->block<3,1>(6, 2) = -Rj.col(0);

        H2->block<3,1>(9, 0) = -Rj.col(1);
        H2->block<3,1>(9, 1) = Rj.col(0);

        // dvec(R_err) / dt = 0
    }

    return residual;
}
